File: src/main/java/golf/tweede/Panama.java
package golf.tweede;

import golf.tweede.gen.java_interop_h;

import org.openjdk.jmh.annotations.Benchmark;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.nio.file.Path;
import java.nio.file.Paths;

public class Panama {
    static {
        Path p = Paths.get("src/main/rust/target/release/libjava_interop.so");
        System.load(p.toAbsolutePath().toString()); // load library
    }

    /**
     * Grab a string from a MemorySegment, and free the original Rust string from
     * the pointer
     * 
     * @param segment which contains a string
     * @return the string
     */
    private static String segmentToString(MemorySegment segment) {
        String string = segment.getString(0);
        java_interop_h.freeString(segment);
        return string;
    }

    public static String doubleToStringRust(double value) {
        return segmentToString(java_interop_h.doubleToStringRust(value));
    }

    public static String doubleToStringRyu(double value) {
        return segmentToString(java_interop_h.doubleToStringRyu(value));
    }

    public static String doubleArrayToStringRyu(double[] array) {
        String output;
        try (Arena offHeap = Arena.ofConfined()) {
            // allocate off-heap memory for input array
            MemorySegment segment = offHeap.allocateFrom(ValueLayout.JAVA_DOUBLE, array);
            output = segmentToString(java_interop_h.doubleArrayToStringRyu(segment, array.length));
        } // release memory for input array

        return output;
    }

    // Benchmarks

    @Benchmark
    public String doubleToStringRustBenchmark(Main.BenchmarkState state) {
        return doubleToStringRust(state.value);
    }

    @Benchmark
    public String doubleToStringRyuBenchmark(Main.BenchmarkState state) {
        return doubleToStringRyu(state.value);
    }

    @Benchmark
    public String doubleArrayToStringRyuBenchmark(Main.BenchmarkState state) {
        return doubleArrayToStringRyu(state.array);
    }
}

File: src/main/java/golf/tweede/Panama.kt
package golf.tweede

import golf.tweede.Main.BenchmarkState
import golf.tweede.gen.java_interop_h
import org.openjdk.jmh.annotations.Benchmark
import java.lang.foreign.Arena
import java.lang.foreign.MemorySegment
import java.lang.foreign.ValueLayout


open class Panama {
    companion object {
        /**
         * Grab a string from a MemorySegment, and free the original Rust string from
         * the pointer
         *
         * @param segment which contains a string
         * @return the string
         */
        private fun segmentToString(segment: MemorySegment): String {
            val string = segment.getString(0)
            java_interop_h.freeString(segment)
            return string
        }

        fun doubleToStringRust(value: Double): String = segmentToString(java_interop_h.doubleToStringRust(value))

        fun doubleToStringRyu(value: Double): String = segmentToString(java_interop_h.doubleToStringRyu(value))

        fun doubleArrayToStringRyu(array: DoubleArray): String {
            val output: String
            Arena.ofConfined().use { offHeap ->
                // allocate off-heap memory for input array
                val segment = offHeap.allocateFrom(ValueLayout.JAVA_DOUBLE, *array)
                output = segmentToString(java_interop_h.doubleArrayToStringRyu(segment, array.size.toLong()))
            }
            return output
        }
    }


    @Benchmark
    fun doubleToStringRustBenchmark(state: BenchmarkState): String {
        return doubleToStringRust(state.value)
    }

    @Benchmark
    fun doubleToStringRyuBenchmark(state: BenchmarkState): String {
        return doubleToStringRyu(state.value)
    }

    @Benchmark
    fun doubleArrayToStringRyuBenchmark(state: BenchmarkState): String {
        return doubleArrayToStringRyu(state.array)
    }
}

fun main() {
    println(Panama.doubleToStringRust(Math.PI))
    println(Panama.doubleToStringRyu(Math.PI))
    println(Panama.doubleArrayToStringRyu(doubleArrayOf(1.0, 2.0, Math.PI)))
}
